File: src/matchlock/dh/commit.ts
import { sha256 } from '@noble/hashes/sha256';
import { bytesToHex, hexToBytes } from '@noble/hashes/utils';
import type { MatchToken, CommitHash } from '../types.js';
import { isValidMatchToken } from '../types.js';
import { InvalidTokenError } from '../errors.js';

export function commitToken(matchToken: MatchToken): CommitHash {
  if (!isValidMatchToken(matchToken)) throw new InvalidTokenError('invalid match token');
  // Hash the 32 raw bytes, not the 64-char hex encoding
  return bytesToHex(sha256(hexToBytes(matchToken))) as CommitHash;
}

export function commitTokens(matchTokens: MatchToken[]): CommitHash[] {
  return matchTokens.map(commitToken);
}

export function verifyCommitment(matchToken: MatchToken, commitHash: CommitHash): boolean {
  const computed = commitToken(matchToken);
  if (computed.length !== commitHash.length) return false;
  let result = 0;
  for (let i = 0; i < computed.length; i++) {
    result |= computed.charCodeAt(i) ^ commitHash.charCodeAt(i);
  }
  return result === 0;
}

File: crates/matchlock-core/src/commit.rs
use sha2::{Sha256, Digest};
use subtle::ConstantTimeEq;
use crate::types::{MatchToken, CommitHash};

pub fn commit_token(token: &MatchToken) -> CommitHash {
    CommitHash(Sha256::digest(token.0).into())
}

pub fn commit_tokens(tokens: &[MatchToken]) -> Vec<CommitHash> {
    tokens.iter().map(commit_token).collect()
}

pub fn verify_commitment(token: &MatchToken, commit_hash: &CommitHash) -> bool {
    bool::from(commit_token(token).0.ct_eq(&commit_hash.0))
}
